src/lobpcg.jl

```diff
@@ -0,0 +1,751 @@
+        bs = iterator.currentBlockSize[]
+        # Update active R and P blocks
+        update_active!(iterator.activeMask, bs, 
+            (iterator.activeRBlocks.block, iterator.RBlocks.block), 
+            (iterator.activePBlocks.block, iterator.PBlocks.block),
+            (iterator.activePBlocks.A_block, iterator.PBlocks.A_block),
+            (iterator.activePBlocks.B_block, iterator.PBlocks.B_block))
+        # Precondition and constrain the active residual vectors
+        precond_constr!(iterator.activeRBlocks.block, bs, iterator.precond!, iterator.constr!)
+        # Orthonormalizes R[:,1:bs] and finds AR[:,1:bs] and BR[:,1:bs]
+        ortho_AB_mul_X!(iterator.activeRBlocks, iterator.ortho!, iterator.A, iterator.B, bs)
+        # Orthonormalizes P and updates AP
+        iterator.ortho!(iterator.activePBlocks, bs, update_AX=true, update_BX=true)
+        # Update the gram matrix [X R P]' A [X R P] and [X R P]' B [X R P]
+        block_grams_3x3!(iterator, bs)
+        # Solve the Rayleigh-Ritz sub-problem
+        sub_problem!(iterator, sizeX, bs, bs)
+        # Updates Ritz vectors X and updates AX and BX accordingly
+        # And updates P, AP and BP
+        update_X_P!(iterator, bs, bs)
+        residuals!(iterator)
+        update_mask!(iterator, residualTolerance)
+    end
+    if log
+        return LOBPCGState(iteration, iterator.residuals[1:sizeX], iterator.ritz_values[1:sizeX])
+    else
+        return LOBPCGState(iteration, nothing, nothing)
+    end
+end
+
+"""
+The Locally Optimal Block Preconditioned Conjugate Gradient Method (LOBPCG)
+
+Finds the k extremal eigenvalues and their corresponding eigenvectors satisfying `AX = λBX` returning `λ` and `X`, using `X` as an initial guess, where `k = size(X,2)`. The input `X` is overwritten with the solution.
+
+`A` and `B` may be generic types but `Base.A_mul_B!(C, AorB, X)` and `AorB * X` must be defined for vectors and strided matrices `X` and `C`.
+
+    lobpcg(A, [B,] largest, X0; kwargs...) -> results
+
+# Arguments
+
+- `A`: linear operator;
+- `B`: linear operator;
+- `largest`: `true` if largest eigenvalues are desired and false if smallest;
+- `X0`: Initial guess, will not be modified. The number of columns is the number of eigenvectors desired.
+
+## Keywords
+
+- `log::Bool`: default is `false`; if `true`, `results.trace` will store iterations    
+    states; if `false` only `results.trace` will be empty;
+
+- `P`: preconditioner of residual vectors, must overload `A_ldiv_B!`;
+
+- `C`: constraint to deflate the residual and solution vectors orthogonal
+    to a subspace; must overload `A_mul_B!`;
+
+- `maxiter`: maximum number of iterations; default is 200;
+
+- `tol::Number`: tolerance to which residual vector norms must be under.
+
+# Output
+
+- `results`: a `LOBPCGResults` struct. `r.λ` and `r.X` store the eigenvalues and eigenvectors.
+
+"""
+function lobpcg(A, largest::Bool, X0::Union{AbstractMatrix, AbstractVector}; kwargs...)
```

Comment:
I think asking the user for an initial guess is a bit awkward; it's probably almost always a random vec or block. Default value would be useful.

Reply:
Sure.

Reply:
I am adding another wrapper function which defaults to `Matrix` and `Float64` and calls `size` on `A`. I think this is a nice compromise since the `X0` functions are still there.

Reply:
Oh `X0` also tells us the number of eigenvectors desired, so that will have to be replaced by `nev` or something..

Reply:
We're advising to wrap the operator with LinearMaps.jl, which implements size and eltype for you.

Reply:
Ok